File: julia/SMC2FC/src/SMC2/MassMatrix.jl
# SMC2/MassMatrix.jl — diagonal-mass-matrix estimation for the per-tempering-
# level HMC kernel.
#
# Direct port of `smc2fc/core/mass_matrix.py`. Comment from Python (kept):
#   "Full mass matrices fail on this problem: the PF likelihood landscape
#   punishes correlated HMC proposals and acceptance collapses to zero by
#   λ ≈ 0.3. The diagonal approximation is stable and adapts per-level
#   from the current particle cloud's per-dimension variance."

module MassMatrix

using Statistics: var

export estimate_mass_matrix

"""
    estimate_mass_matrix(particles::AbstractMatrix; regularisation=1e-4) -> Vector

Diagonal inverse mass-matrix from per-dimension particle variance.
Returns a `Vector{T}` of length `d_theta` (AdvancedHMC's `DiagEuclideanMetric`
expects this layout). Variance is floored at `regularisation` to avoid
zero-mass dimensions.
"""
function estimate_mass_matrix(particles::AbstractMatrix{T};
                               regularisation::Real = 1e-4) where {T<:Real}
    v = vec(var(particles; dims=1))
    return max.(v, T(regularisation))
end

end # module MassMatrix

File: smc2fc_JAX_functional/smc2fc/core/mass_matrix.py
"""Diagonal mass-matrix estimation for the per-tempering-level HMC kernel.

Full mass matrices fail on this problem: the PF likelihood landscape
punishes correlated HMC proposals and acceptance collapses to zero by
λ ~ 0.3. The diagonal approximation is stable and adapts per-level
from the current particle cloud's per-dimension variance.
"""

import jax.numpy as jnp
from jax import Array


def estimate_mass_matrix(particles: Array,
                          regularisation: float = 1e-4) -> Array:
    """Returns the diagonal inverse mass matrix from per-dim variance.

    Pure JAX: variance is computed across the particle axis, then
    floored at ``regularisation`` to keep HMC stable when a dim
    happens to be (near-)deterministic.

    Args:
        particles: Particle cloud of shape ``(n_smc, n_dim)``.
        regularisation: Lower bound applied to each variance entry.
            Prevents the mass matrix from blowing up when a dim has
            collapsed to a near-point estimate.

    Returns:
        Diagonal inverse mass matrix of shape ``(1, n_dim)``. The
        leading ``1`` axis matches BlackJAX HMC's expected
        ``inverse_mass_matrix`` shape (broadcast over chains).
    """
    var = jnp.var(particles, axis=0)
    var = jnp.maximum(var, regularisation)
    return var[None, :]
